pkg/build/types/parser/deffile_test.go

```diff
@@ -148,3 +235,144 @@ func TestParseDefinitionFileFailure(t *testing.T) {
+		t.Fatal("unexpecter result from Error()", myValidErr1.Error())
+	}
+}
+
+// Specific tests to cover some corner cases of PopulateDefinition()
+func TestPopulateDefinition(t *testing.T) {
+	//
+	// Some variables used throughout the tests
+	//
+
+	// We use a specific set of section names to reach some corner cases
+	testMap := make(map[string]string)
+	testMap["files"] = "file1 file2"
+	testMap["labels"] = "label1"
+
+	emptyMap := make(map[string]string)
+
+	//
+	// Test with invalid data
+	//
+	invalidData := new(types.Definition)
+	invalidData.Labels = make(map[string]string)
+	populateDefinition(emptyMap, invalidData)
+
+	//
+	// Test with very specific maps
+	//
+
+	// A structure to store results (not really relevant here)
+	myData := new(types.Definition)
+	myData.Labels = make(map[string]string)
+
+	myerr := populateDefinition(testMap, myData)
+	if myerr != nil {
+		t.Fatal("Test failed while testing populateDefinition()")
+	}
+}
+
+// Specific tests to cover some corners cases of doHeader()
+func TestDoHeader(t *testing.T) {
+	invalidHeader := "headerTest"
+	myData := new(types.Definition)
+	myData.Labels = make(map[string]string)
+
+	myerr := doHeader(invalidHeader, myData)
+	if myerr == nil {
+		t.Fatal("Test succeeded while supposed to fail")
+	}
+}
+
+func TestIsValidDefinition(t *testing.T) {
+
+	//
+	// Test with a bunch of valid files
+	//
+	valid_tests := []struct {
+		name     string
+		defPath  string
+		sections string
+	}{
+		{"Arch", "testdata_good/arch/arch", "testdata_good/arch/arch_sections.json"},
+		{"BusyBox", "testdata_good/busybox/busybox", "testdata_good/busybox/busybox_sections.json"},
+		{"Debootstrap", "testdata_good/debootstrap/debootstrap", "testdata_good/debootstrap/debootstrap_sections.json"},
+		{"Docker", "testdata_good/docker/docker", "testdata_good/docker/docker_sections.json"},
+		{"LocalImage", "testdata_good/localimage/localimage", "testdata_good/localimage/localimage_sections.json"},
+		{"Scratch", "testdata_good/scratch/scratch", "testdata_good/scratch/scratch_sections.json"},
+		{"Shub", "testdata_good/shub/shub", "testdata_good/shub/shub_sections.json"},
+		{"Yum", "testdata_good/yum/yum", "testdata_good/yum/yum_sections.json"},
+		{"Zypper", "testdata_good/zypper/zypper", "testdata_good/zypper/zypper_sections.json"},
+	}
+
+	for _, tt := range valid_tests {
+		t.Run(tt.name, test.WithoutPrivilege(func(t *testing.T) {
+			deffile := tt.defPath
+
+			valid, err := IsValidDefinition(deffile)
+			if valid == false || err != nil {
+				t.Fatal("Validation of a definition file failed while expected to succeed")
+			}
+		}))
+	}
+
+	//
+	// Test with a non-existing file
+	//
+	valid, err := IsValidDefinition("notExistingDirectory/notExistingFile")
+	if valid == true && err != nil {
+		t.Fatal("Validation of a non-existing file succeeded while expected to fail")
+	}
+
+	//
+	// Test passing a valid directory in instead of a file
+	//
+	valid, err = IsValidDefinition("testdata_bad")
```

Comment:
Looks like there is a file descriptor leak in `IsValidDefinition` otherwise `defer defFile.Close()` is not reached when definition is a directory